### PR TITLE
Wip/6.0 scrollablefetching dynamicbatch

### PR DESCRIPTION
**Pull request by dreab8**
this PR :

- implement fetch scroll
- fix unnecessary query for an optional OneToOne with Bytecode enhancement
- lazy loading of lazy basic attributes when an entity is fetched
- dynamic batch

### CODE REVIEW

**Review by sebersole — COMMENTED** [review]
lgtm

**Review by beikov — CHANGES_REQUESTED** [review]
submitting in-progress review

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/FetchParentAccess.java [hunk 1]
```diff
@@ -26,6 +27,8 @@
 
 	NavigablePath getNavigablePath();
 
+	EntityKey getEntityKey();
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/sql/results/graph/FetchParentAccess.java:30]: I'll do thanks

File: hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java [hunk 2]
```diff
@@ -41,88 +46,142 @@ public FetchingScrollableResultsImpl(
...
+	private static <R> EntityInitializer extractResultInitializer(RowReader<R> rowReader) {
+		for ( Initializer initializer : rowReader.getInitializers() ) {
+			if ( initializer instanceof EntityInitializer ) {
+				EntityInitializer entityInitializer = (EntityInitializer) initializer;
+				if ( entityInitializer instanceof EntityResultInitializer ) {
+					return entityInitializer;
+				}
+			}
+		}
+		return null;
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java:64]: you are right, I'm changing it, Thanks!

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/FetchParentAccess.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/sql/results/graph/FetchParentAccess.java:30]: yeap

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/embeddable/internal/EmbeddableResultInitializer.java [hunk 3]
```diff
@@ -22,7 +23,12 @@ public EmbeddableResultInitializer(
 
 	@Override
 	public Object getParentKey() {
-		return null;
+		return findFirstEntityDescriptorAccess().getParentKey();
+	}
+
+	@Override
+	public EntityKey getEntityKey() {
+		return findFirstEntityDescriptorAccess().getEntityKey();
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/sql/results/graph/embeddable/internal/EmbeddableResultInitializer.java:31]: :+1:

File: hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java [hunk 4]
```diff
@@ -263,16 +310,60 @@ else if ( maxPosition != null && rowNumber == maxPosition ) {
...
+		final RowReader<R> rowReader = getRowReader();
+
+		boolean afterLast = false;
+		boolean resultProcessed = false;
+
+		final EntityKey entityKey = getEntityKey();
+
+		currentRow = rowReader.readRow( getRowProcessingState(), getProcessingOptions() );
+
+//		currentRow = new Object[] {row};
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java:327]: :+1:

File: hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java [hunk 5]
```diff
@@ -41,88 +46,142 @@ public FetchingScrollableResultsImpl(
...
+				keyToRead = getEntityKey();
+			}
+			else {
+				// Since the result set cursor is always left at the first
+				// physical row after the "last processed", we need to jump
+				// back one position to get the key value we are interested
+				// in skipping
+
+				getRowProcessingState().previous();
+//				getRowProcessingState().previous();
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java:141]: :+1:

File: hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java [hunk 6]
```diff
@@ -41,88 +46,142 @@ public FetchingScrollableResultsImpl(
...
+			// thus if we are after the last physical row, this might be
+			// caused by either:
+			//      1) scrolling to the last logical row
+			//      2) scrolling past the last logical row
+			// In the latter scenario, the previous logical row
+			// really is the last logical row.
+			//
+			// In all other cases, we should process back two
+			// logical records (the current logic row, plus the
+			// previous logical row).
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/FetchingScrollableResultsImpl.java:128]: :+1:

**Pull request merged**